float8_experimental/float8_ops.py

```diff
@@ -78,9 +79,16 @@ def preprocess_addmm(a: Float8Tensor, b: Float8Tensor):
 
 @implements([aten.mm.default])
 def float8_mm(aten_op, args, kwargs=None):
-    assert isinstance(args[0], Float8Tensor) and isinstance(args[1], Float8Tensor)
     a = args[0]
     b = args[1]
+    if isinstance(b, AsyncCollectiveTensor):
```

Comment:
not sure how this is creeping in here

Reply:
this is a bit surprising, so this change fixed the AsyncCollectiveTensor issue?

Reply:
Yeah, exactly, Im not totally sure yet how this collective made it into the dispatch  for Float8Tensor

Reply:
yeah tbh this is the only part I don't love I am having a hard time figure out where this is creeping in.. 

This dispatch should be DTensor(Float8Tensor()) -> aten.mm.defualt
.... okay as I am writing this I think I migh tnow


I think we have 
```
DTensor(Float8Tensor)), DTensor(AsyncTensor(Float8Tensor) -> aten.mm.defualt
```

The first layer of torch.dispatch goes through Dtensors table for both. Then we have unwrap
 
 ```
 Float8Tensor(), Async(Float8Tensor) -> aten.mm.default
 ```
 
 And the argparse chooses the leftmost subclass to dispatch too and thus we need this. I wonder if we can either handle this in DTensor or if there is another way?
 
 cc @bdhirsh  if this makes sense as well

Reply:
If your Float8Tensor class expects any other subclass to run before you, you can add this to your class:
```python
        if not all(issubclass(cls, t) for t in types):
            return NotImplemented
```

The NotImplemented will trigger the handler of the next type (just like it does for `__add__` in regular python)

Reply:
Thanks Alban this makes alot  of sense, just updated and everything appears to be working as expected!